sklearn/datasets/openml.py

```diff
@@ -86,10 +88,39 @@ def is_gzip(_fsrc):
     req = Request(_OPENML_PREFIX + openml_path)
     req.add_header('Accept-encoding', 'gzip')
 
+    def _md5_validated_stream(input_stream, md5_checksum):
+        """
+        Consume binary stream to validate checksum,
+        return a new stream with same content
+
+        Parameters
+        ----------
+        input_stream : io.BufferedIOBase
+            Input stream with a read() method to get content in bytes
+
+        md5_checksum: str
+            Expected md5 checksum
+
+        Returns
+        -------
+        BytesIO stream with the same content as input_stream for consumption
+        """
+        with closing(input_stream):
+            bytes_content = input_stream.read()
```

Comment:
Can we do this without consume the whole stream? https://docs.python.org/3/library/hashlib.html?highlight=hashlib#examples

Reply:
I don't think so. I could only think of this: to verify the hash we need to read the stream, so we read it and then return a new stream to keep the methods API consistent.

Did you mean consuming the stream in chunks and updating the hashlib.md5 instance with the chunks instead of reading it all in one go? I don't think that would make any difference.

Reply:
Yes I mean in chunks.

Currently, with this PR the stream will be read twice, once during the the md5 check and again by the caller of `_open_openml_url`.

Another option is to have `_open_openml_url` return the byte content such that the callers do not need to call read.

Reply:
Done the second option, now reads stream once and passes content along